File: sim_1_6_5.c
/* program that generates input for 1.6.5.sol.*
 */
#include <stdio.h>
#include <time.h>
#include <unistd.h> /* optarg */
#include <stdlib.h> /* RAND_MAX */
#include "defs.h"


void c_clear(FILE* file, int columns, int rows)
{
    fprintf(file, "C\n");
}


void c_color(FILE* file, int columns, int rows)
{
    fprintf(file, "L %d %d %c\n",
            rand()%columns+1, rand()%rows+1, rand()%26+'A');
}


void c_range(int max, int* low, int* high)
{
    *low = rand() % (max-1) + 1;
    *high = rand() % (max-*low) + *low+1;
}


void c_vertical(FILE* file, int columns, int rows)
{
    int low, high;
    c_range(rows, &low, &high);
    fprintf(file, "V %d %d %d %c\n", rand()%columns+1, low, high, rand()%26+'A');
}


void c_horizontal(FILE* file, int columns, int rows)
{
    int low, high;
    c_range(columns, &low, &high);
    fprintf(file, "H %d %d %d %c\n", low, high, rand()%rows+1, rand()%26+'A');
}


void c_rectangle(FILE* file, int columns, int rows)
{
    int low_column, low_row, high_column, high_row;
    c_range(columns, &low_column, &high_column);
    c_range(rows, &low_row, &high_row);
    fprintf(file, "K %d %d %d %d %c\n", low_column, low_row,
            high_column, high_row, rand()%26+'A');
}


void c_region(FILE* file, int columns, int rows)
{
    fprintf(file, "F %d %d %c\n",
            rand()%columns+1, rand()%rows+1, rand()%26+'A');
}


#define FILE_SAVE "1.6.5.rand"
void c_save(FILE* file, int columns, int rows)
{
    fprintf(file, "S %s\n", FILE_SAVE);
}


#define MAX_COMMANDS 7
void generate(int num_cases, FILE* file, int columns, int rows)
{
    /* define and initialize an array of pointer to prints */
    static void (*commands[MAX_COMMANDS])(FILE*, int, int);
    static int init = 0;

    if (init == 0)
    {
        commands[0] = &c_clear;
        commands[1] = &c_color;
        commands[2] = &c_vertical;
        commands[3] = &c_horizontal;
        commands[4] = &c_rectangle;
        commands[5] = &c_region;
        commands[6] = &c_save;

        init = 1;
    }

    int i;
    for (i = 0; i < num_cases; i++)
    {
        int index = rand() % MAX_COMMANDS;
        (*commands[index])(file, columns, rows);
    }
}



void main(int argc, char** argv)
{
    char opt;
    int num_cases = 0;
    int max_rows = 0;
    int max_columns = 0;
    char* output_file = NULL;

    while ((opt = getopt(argc, argv, "c:r:l:o:")) != -1)
    {
        switch (opt)
        {
        case 'c':
            num_cases = atoi(optarg);
            break;
        case 'r':
            max_rows = atoi(optarg);
            break;
        case 'l':
            max_columns = atoi(optarg);
            break;
        case 'o':
            output_file = optarg;
            break;
        default:
            printf("Usage: %s -c num_cases -r max_rows -l max_columns -o output_file\n",
                   argv[0]);
        }
    }

    if (num_cases == 0 || max_rows == 0 || max_columns == 0 ||
        output_file == NULL)
        printf("Usage: %s -c num_cases -r max_rows -l max_columns -o output_file\n",
               argv[0]);
    else
    {
        /* must make the total number of cases equal (MAX_CASES-3 -- to
         * accommodate the 3 commands I, S and X), so they may altogether fit
         * the array of commands declared in the various solutions to this
         * problem
         */
        if (num_cases > MAX_CASES-3)
            num_cases = MAX_CASES-3;


        /* begin the session */
        srand(time(NULL));
        FILE* file = fopen(output_file, "w");
        int rows = rand()%max_rows+1;
        int columns = rand()%max_columns+1;
        fprintf(file, "I %d %d\n", columns, rows);


        /* generate a bunch of commands */
        generate(num_cases, file, columns, rows);


        /* make at least one save */
        fprintf(file, "S %s\n", FILE_SAVE);


        /* terminate the session */
        fprintf(file, "X\n");
        fclose(file);
    }
}

File: sim_1_6_5.java
import java.io.File;
import java.io.PrintWriter;
import java.io.FileNotFoundException;
import java.util.Random;
import org.apache.commons.cli.*;


class sim_1_6_5 {
    static PrintWriter writer;
    static Random random;


    interface Functor {
        public void execute(int columns, int rows);
    }


    static class FClear implements Functor {
        public void execute(int columns, int rows) {
            writer.write("C\n");
            System.out.print("C\n");
        }
    }


    static class FColor implements Functor {
        public void execute(int columns, int rows) {
            String line = String.format("L %d %d %c\n",
                                        random.nextInt(columns)+1,
                                        random.nextInt(rows)+1,
                                        random.nextInt(26)+'A');
            writer.write(line);
            System.out.print(line);
        }
    }


    static class Range {
        int _low;
        int _high;

        Range(int max) {
            _low = random.nextInt(max-1) + 1;
            _high = random.nextInt(max-_low) + _low+1;
        }
    }


    static class FVertical implements Functor {
        public void execute(int columns, int rows) {
            Range range = new Range(rows);
            String line = String.format("V %d %d %d %c\n",
                                        random.nextInt(columns)+1,
                                        range._low, range._high,
                                        random.nextInt(26)+'A');
            writer.write(line);
            System.out.print(line);
        }
    }


    static class FHorizontal implements Functor {
        public void execute(int columns, int rows) {
            Range range = new Range(columns);
            String line = String.format("H %d %d %d %c\n",
                                        range._low, range._high,
                                        random.nextInt(rows)+1,
                                        random.nextInt(26)+'A');
            writer.write(line);
            System.out.print(line);
        }
    }


    static class FRectangle implements Functor {
        public void execute(int columns, int rows) {
            Range range1 = new Range(columns);
            Range range2 = new Range(rows);
            String line = String.format("K %d %d %d %d %c\n",
                                        range1._low, range2._low,
                                        range1._high, range2._high,
                                        random.nextInt(26)+'A');
            writer.write(line);
            System.out.print(line);
        }
    }


    static class FRegion implements Functor {
        public void execute(int columns, int rows) {
            String line = String.format("F %d %d %c\n",
                                        random.nextInt(columns)+1,
                                        random.nextInt(rows)+1,
                                        random.nextInt(26)+'A');
            writer.write(line);
            System.out.print(line);
        }
    }


    static class FSave implements Functor {
        public void execute(int columns, int rows) {
            String line = String.format("S %s\n", FILE_SAVE);
            writer.write(line);
            System.out.print(line);
        }
    }


    private static final int MAX_COMMANDS = 7;
    static void generate(int num_cases, int columns, int rows) {
        Functor[] functors = null;

        if (functors == null)
        {
            functors = new Functor[MAX_COMMANDS];

            functors[0] = new FClear();
            functors[1] = new FColor();
            functors[2] = new FVertical();
            functors[3] = new FHorizontal();
            functors[4] = new FRectangle();
            functors[5] = new FRegion();
            functors[6] = new FSave();
        }

        for (int i = 0; i < num_cases; i++)
        {
            int index = random.nextInt(MAX_COMMANDS);
            functors[index].execute(columns, rows);
        }
    }


    private static final String FILE_SAVE = "tmp_1_6_5";
    public static void main(String[] args) throws ParseException, FileNotFoundException {
        Options options = new Options();
        options.addOption("c", true, "Number of cases");
        options.addOption("r", true, "Maximum rows");
        options.addOption("l", true, "Maximum columns");
        options.addOption("d", true, "Maximum commands");
        options.addOption("o", true, "Output filename");


        CommandLineParser parser = new PosixParser();
        CommandLine cmd = parser.parse(options, args);


        int num_cases = 0;
        int max_rows = 0;
        int max_columns = 0;
        int max_commands = 0;
        String output_file = null;


        if (cmd.hasOption("c"))
            num_cases = Integer.parseInt(cmd.getOptionValue("c"));
        if (cmd.hasOption("r"))
            max_rows = Integer.parseInt(cmd.getOptionValue("r"));
        if (cmd.hasOption("l"))
            max_columns = Integer.parseInt(cmd.getOptionValue("l"));
        if (cmd.hasOption("o"))
            output_file = cmd.getOptionValue("o");


        // begin the session
        writer = new PrintWriter(new File(output_file));
        random = new Random();
        int rows = random.nextInt(max_rows)+2;
        int columns = random.nextInt(max_columns)+2;
        String line = String.format("I %d %d\n", columns, rows);
        writer.write(line);
        System.out.print(line);


        // generate a bunch of commands
        generate(num_cases, columns, rows);


        // make at least one save
        line = String.format("S %s\n", FILE_SAVE);
        writer.write(line);
        System.out.print(line);


        // terminate the session
        writer.write("X\n");
        System.out.print("X\n");


        writer.close();
    }
}
